File: Leetcode/572_Subtree_of_Another_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} s
 * @param {TreeNode} t
 * @return {boolean}
 */
var isSubtree = function(s, t) {
    if (s) {
        if (isSame(s, t)) {
            return true;
        }
        return isSubtree(s.left, t) || isSubtree(s.right, t);
    }
    return false;
};

var isSame = function(root, newRoot) {
    if (root === null && newRoot === null) {
        return true;
    }
    if (root !== null && newRoot !== null) {
        if (root.val === newRoot.val) {
            return isSame(root.left, newRoot.left) && isSame(root.right, newRoot.right);
        }
        return false;
    }
    
    return false;
}

File: Leetcode/0572/572_Subtree_of_Another_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSubtree(TreeNode* s, TreeNode* t) {
        if(s == nullptr)
        {
            return false;
        }
        if(isSame(s, t))
        {
            return true;
        }
        
        return isSubtree(s->left, t) || isSubtree(s->right, t);
    }
    
    bool isSame(TreeNode* s, TreeNode* t)
    {
        if(s == nullptr && t == nullptr)
        {
            return true;
        }
        if(s == nullptr || t == nullptr)
        {
            return false;
        }
        
        if(s->val != t->val)
        {
            return false;
        }
        
        return isSame(s->left, t->left) && isSame(s->right, t->right);
    }
};